### ISSUE TITLE
Large cumulative sums appear to be nondeterministic. 

### ISSUE THREAD

**Issue by tom-p-reichel** [pytorch/pytorch#75240]
### 🐛 Describe the bug

@timothytmzhou and I found that cumulative sums over large input arrays do not seem to deterministically produce the same output array:

```python
import torch
import math
import random
import numpy as np
import os

# make ourselves as deterministic as possible.
os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
np.random.seed(0)
torch.manual_seed(0)
random.seed(0)
torch.use_deterministic_algorithms(True)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

# (deterministic) messy number generator
v = torch.Tensor([(x*math.pi)%1 for x in range(1000000)]).to(torch.device("cuda"))

# init a and b to the same cumulative sum.
a = b = v.cumsum(0)
counter = 0

# this loop should never terminate.
while ((a==b).all()):
	b = v.cumsum(0)
	counter+=1

print(f"non-determinism after {counter} runs.")
print("errors:",(a-b).abs()[a-b!=0])
print("first error index:",torch.where((a-b)!=0)[0][0])
```
yields, on my machine:
```
non-determinism after 1 runs.
errors: tensor([0.0020, 0.0020, 0.0020,  ..., 0.2188, 0.2188, 0.2188], device='cuda:0')
first error index: tensor(38400, device='cuda:0')
```
Multiple runs of this file exhibit unique non-determinism. Here's another run:
```
non-determinism after 1 runs.
errors: tensor([0.0039, 0.0039, 0.0039,  ..., 0.0938, 0.0938, 0.0938], device='cuda:0')
first error index: tensor(76800, device='cuda:0')
```
We decided to test this after reading [this paper](https://www.cs.purdue.edu/homes/lintan/publications/variance-ase20.pdf). Of note is the following paragraph:

> Scheduling and floating-point precision: GPU programming uses warp as a unit of computation. A warp consists of 32 parallel threads with concurrent memory access. Due to the limited precision (32-bits) of DL models, rounding errors are introduced at every step of floating-point calculation. In the GPU model, concurrent accesses to a single memory address must be serialized to prevent race conditions with no guaranteed order of access. Therefore, the rounding error introduced in each warp may vary across fixed-seed identical training runs due to the different accessing orders.

### Versions

Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 11.2.0
Clang version: 13.0.1
CMake version: version 3.22.3
Libc version: glibc-2.35

Python version: 3.10.4 (main, Mar 23 2022, 23:05:40) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.16.16-arch1-1-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.112
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1050 Ti
Nvidia driver version: 510.54
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.3.2
/usr/lib/libcudnn_adv_infer.so.8.3.2
/usr/lib/libcudnn_adv_train.so.8.3.2
/usr/lib/libcudnn_cnn_infer.so.8.3.2
/usr/lib/libcudnn_cnn_train.so.8.3.2
/usr/lib/libcudnn_ops_infer.so.8.3.2
/usr/lib/libcudnn_ops_train.so.8.3.2
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] torch==1.11.0
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @ngimel @mruberry @kurtamohler

**Comment by ezyang** [comment]
cc @kurtamohler

**Comment by ezyang** [comment]
This repros for me

```
(/private/home/ezyang/pytorch-tmp-env) ezyang@devfair040:~/pytorch-tmp$ python repro.py
non-determinism after 1 runs.
errors: tensor([0.0010, 0.0010, 0.0010,  ..., 0.0938, 0.0938, 0.0938], device='cuda:0')
first error index: tensor(28800, device='cuda:0')
(/private/home/ezyang/pytorch-tmp-env) ezyang@devfair040:~/pytorch-tmp$ nvidia-smi
Mon Apr  4 18:39:57 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.57.02    Driver Version: 470.57.02    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Quadro GP100        On   | 00000000:81:00.0 Off |                    0 |
| 26%   33C    P0    31W / 235W |      1MiB / 16278MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```

**Comment by ezyang** [comment]
CUDA cumsum uses CUB's InclusiveScan which is supposed to be deterministic per docs here https://nvlabs.github.io/cub/structcub_1_1_device_scan.html#a16d7bc049ba8985dd89b10b3bcf0a8a3

**Comment by ngimel** [comment]
Pytorch directly calls thrust in this case, cc @zasdfgbnm (this is small enough to use int32 indexing, so the issue is not multiple cub kernels)

**Comment by ezyang** [comment]
Am I misunderstanding the profile? Lol

```
==2218205== Profiling result:                                                                                                              Type  Time(%)      Time     Calls       Avg       Min       Max  Name                                               GPU activities:   49.87%  429.40us         1  429.40us  429.40us  429.40us  [CUDA memcpy HtoD]                                                    8.36%  71.936us         3  23.978us  23.168us  24.768us  void at::native::vectorized_elementwise_kernel<int
=4, at::native::CUDAFunctor_add<float>, at::detail::Array<char*, int=3>>(int, float, at::native::CUDAFunctor_add<float>)                           5.76%  49.568us         3  16.522us  5.3760us  22.368us  void at_cuda_detail::cub::DeviceSelectSweepKernel<at_cuda_detail::cub::DispatchSelectIf<at_cuda_detail::cub::CountingInputIterator<long, long>, at_cuda_detail::cub::TransformInp
utIterator<bool, at::native::_GLOBAL__N__42_tmpxft_0020afff_00000000_7_Nonzero_cpp1_ii_cba1aaa0::NonZeroOp<bool>, bool*, long>, long*, int*, at_cuda_detail::cub::NullType, at_cuda_detail::cub::NullType, int, bool=0>::PtxSelectIfPolicyT, at_cuda_detail::cub::CountingInputIterator<long, long>, at_cuda_detail::cub::TransformInputIterator<bool, at::native::_GLOBAL__N__42_tmpxft_0020
afff_00000000_7_Nonzero_cpp1_ii_cba1aaa0::NonZeroOp<bool>, bool*, long>, long*, int*, at_cuda_detail::cub::ScanTileState<int, bool=1>, at_cuda_detail::cub::NullType, at_cuda_detail::cub::NullType, int, bool=0>(long, at_cuda_detail::cub::CountingInputIter
ator<long, long>, bool, bool, at::native::_GLOBAL__N__42_tmpxft_0020afff_00000000_7_Nonzero_cpp1_ii_cba1aaa0::NonZeroOp<bool>, 
bool*, long, at_cuda_detail::cub::TransformInputIterator<bool, at::native::_GLOBAL__N__42_tmpxft_0020afff_00000000_7_Nonzero_cp
p1_ii_cba1aaa0::NonZeroOp<bool>, bool*, long>, int)                                                                            
                    4.61%  39.711us         2  19.855us  18.783us  20.928us  void at_cuda_detail::cub::DeviceScanKernel<at_cuda
_detail::cub::AgentScanPolicy<int=128, int=15, float, at_cuda_detail::cub::BlockLoadAlgorithm, at_cuda_detail::cub::CacheLoadMo
difier, at_cuda_detail::cub::BlockStoreAlgorithm, at_cuda_detail::cub::BlockScanAlgorithm, at_cuda_detail::cub::MemBoundScaling
<int=128, int=15, float>>, float*, float*, at_cuda_detail::cub::ScanTileState<float, bool=1>, std::plus<float>, at_cuda_detail:
:cub::NullType, int>(int=15, float, at_cuda_detail::cub::BlockLoadAlgorithm, int, at_cuda_detail::cub::CacheLoadModifier, at_cu
da_detail::cub::BlockStoreAlgorithm, at_cuda_detail::cub::BlockScanAlgorithm)                                                  
                    4.44%  38.208us         4  9.5520us  2.6560us  22.080us  void at::native::vectorized_elementwise_kernel<int
=4, at::native::BinaryFunctor<float, float, bool, at::native::_GLOBAL__N__50_tmpxft_0020a9e4_00000000_7_CompareEQKernel_cpp1_ii
_d8008c96::CompareEqFunctor<float>>, at::detail::Array<char*, int=3>>(int, float, float)                                       
                    3.65%  31.392us         2  15.696us  14.944us  16.448us  _ZN2at6native13reduce_kernelILi512ELi1ENS0_8Reduce
OpIbNS0_14func_wrapper_tIbZZZNS0_15and_kernel_cudaERNS_14TensorIteratorEENKUlvE_clEvENKUlvE22_clEvEUlbbE_EEjbLi4EEEEEvT1_                          3.34%  28.800us         4  7.2000us  2.5280us  11.264us  void at::native::vectorized_elementwise_kernel<int
=4, at::native::BUnaryFunctor<float, float, bool, at::native::_GLOBAL__N__50_tmpxft_0020a9e4_00000000_7_CompareEQKernel_cpp1_ii
_d8008c96::CompareEqFunctor<float>>, at::detail::Array<char*, int=2>>(int, float, float)
```

**Comment by ngimel** [comment]
My profile (after removing all prints and comparisons):
```
(pytorch) [ngimel@devgpu006.prn1 ~/local/playground] nvprof --print-gpu-trace python nondet-cumsum.py 
==2837118== NVPROF is profiling process 2837118, command: python nondet-cumsum.py
torch.Size([1000000])
==2837118== Profiling application: python nondet-cumsum.py
==2837118== Profiling result:
   Start  Duration            Grid Size      Block Size     Regs*    SSMem*    DSMem*      Size  Throughput  SrcMemType  DstMemType           Device   Context    Stream  Name
2.23178s  705.15us                    -               -         -         -         -  3.8147MB  5.2830GB/s    Pageable      Device  Tesla V100-SXM2         1         7  [CUDA memcpy HtoD]
2.23335s  4.5120us              (5 1 1)       (128 1 1)        16        0B        0B         -           -           -           -  Tesla V100-SXM2         1         7  void at_cuda_detail::cub::DeviceScanInitKernel<at_cuda_detail::cub::ScanTileState<float, bool=1>>(float, int) [984]
2.23337s  15.264us            (521 1 1)       (128 1 1)        48  7.5156KB        0B         -           -           -           -  Tesla V100-SXM2         1         7  void at_cuda_detail::cub::DeviceScanKernel<at_cuda_detail::cub::DeviceScanPolicy<float>::Policy600, float*, float*, at_cuda_detail::cub::ScanTileState<float, bool=1>, std::plus<float>, at_cuda_detail::cub::NullType, int>(at_cuda_detail::cub::DeviceScanPolicy<float>::Policy600, float*, float*, int, float, bool=1, at_cuda_detail::cub::ScanTileState<float, bool=1>) [993]
```
1st cub kernel is setting up some configs, second cub kernel is doing the work.

**Comment by zasdfgbnm** [comment]
I also reproduced. Will try to create a standalone C++ non-pytorch repro.

**Comment by ngimel** [comment]
The first mismatch is typically at index 76800 on my machine, so inputs smaller than 76800 don't have this problem, inputs larger than 76800 do.

**Comment by zasdfgbnm** [comment]
On my machine, it was 69120.

**Comment by ngimel** [comment]
An on colab with T4 I just got 63360

**Comment by zasdfgbnm** [comment]
Reproduced by hijacking cub's unit test: https://github.com/NVIDIA/cub/pull/453

**Comment by zasdfgbnm** [comment]
cc: @ptrblck

**Comment by tlringer** [comment]
@ezyang : @tom-p-reichel is looking to talk to someone about adding some determinism testing to Pytorch's CI as well, to catch issues like this even when they're from upstream dependencies like CUDA. Is there anyone he can talk to about this? I can set up a meeting if it helps. 

(Everyone else, hi, I'm Talia and I'm working with Tom and Timothy on a research project! Thank you for being so prompt about this.)

**Comment by ezyang** [comment]
Basically everyone relevant is CC'ed to this issue :) What's the proposal? One problem is we don't want to slow down the CI too much, but a lot of these determinism issues only manifest on big inputs (as is the case here!) Maybe we can set some sort of offline fuzzing style infra. (cc @malfet, maybe?)

**Comment by malfet** [comment]
We can add determinism tests on say periodic/nightly basis. fuzzing might lead to increased test flakiness, which is not a desired outcome.

**Comment by ezyang** [comment]
Oh 20sec is totally fine. If you want you can just submit a Pr to add this as a normal test

**Comment by kennethchiu** [comment]
> CUDA cumsum uses CUB's InclusiveScan which is supposed to be deterministic per docs here https://nvlabs.github.io/cub/structcub_1_1_device_scan.html#a16d7bc049ba8985dd89b10b3bcf0a8a3

For what it's worth, the docs reference this paper (https://research.nvidia.com/sites/default/files/pubs/2016-03_Single-pass-Parallel-Prefix/nvr-2016-002.pdf), which actually seems to be saying that the algorithm is non-deterministic, and thus the paper from which the code was presumably based contradicts the docs.

> Another important distinction between our method and prior
work is nondeterministic execution scheduling. Whereas
contemporary scan parallelizations are constructed from static
data flow networks (i.e., the order of operations is fixed for a
given problem setting), our method allows parallel processors to
perform redundant work as necessary to avoid delays from serial
dependences. An interesting implication is that scan results are
not necessarily deterministic for pseudo-associative operators.
For example, the prefix sum across a given floating point dataset
may vary from run to run because the number and order of scan
operators applied by CUB may vary from one run to the next.
Our CUDA-based implementation is freely available within
the open-source CUB library of GPU parallel primitives (v1.0.1
and later) [21].

**Comment by tlringer** [comment]
From looking at that paper, those docs, and the Nvidia comments, it seems then like they intended to introduce this nondeterminism, but did not update the docs or tell downstream consumers?

I wonder then if there will be any way to enforce determinism when it is desired? In the meantime, maybe it's worth documenting in Pytorch on the page about reproducibility that these functions introduce nondeterminism.

Someone wonderful on Twitter tried this on TPUs out of curiosity, and it was not an issue with TPUs.

**Comment by tlringer** [comment]
One more thing: I've been looking around the repo for other uses of InclusiveScan, cumsum, cumprod (is cumprod a problem, @tom-p-reichel ?) , and so on. I find a lot of hits, but I don't know the code base well. I'm curious what parts of Pytorch this might impact downstream. Like anything that might depend on the nondeterministic code. Does anyone know?

**Comment by tlringer** [comment]
(The reason I ask about the above is that it appears that, with identical loss, on fixed hardware, with a fixed random seed, with everything set to be as deterministic as possible, we are getting different gradients---@timothytmzhou is looking into this more to determine what is happening, but it is really difficult to debug the code. It seems to happen after the call to `backward`, which looks to be implemented in C++, and I'm not sure what it does or what it calls.)

**Comment by ngimel** [comment]
@tlringer have you tried [set_deterministic_debug_mode](https://pytorch.org/docs/master/generated/torch.set_deterministic_debug_mode.html?highlight=set_deterministic#torch.set_deterministic_debug_mode)? It won't warn on cumsum yet (cc @kurtamohler to add the warning) because we didn't realize it's nondeterministic, but it should flag other operations known to be non-deterministic.

**Comment by tesfaldet** [comment]
Sounds related to what I experienced with [checkpoint_sequential](https://twitter.com/mattierialgirl/status/1512483082309451782?s=20) the other day.

**Comment by tlringer** [comment]
Thanks, @ngimel. Setting that mode did not flag anything. I'll let you know if learn anything else about what is happening and whether or not it is related.

**Comment by sylam** [comment]
Any plans to make cuda cumsum deterministic for large array sizes/

**Comment by leo-liuzy** [comment]
Any update on plans to make cuda cumsum deterministic for large array sizes?

**Comment by kurtamohler** [comment]
Reopening because of https://github.com/pytorch/pytorch/pull/136224#discussion_r1799866598

**Comment by 0ximjosh** [comment]
Why was this closed? Is this a nofix, or was a fix implemented for this, and if so what patch?

**Comment by ngimel** [comment]
#140887

**Comment by rasbt** [comment]
Just skimming through it, it looks like a fix! Many thanks!!
